src/mem_table_flush.rs

```diff
@@ -73,14 +73,18 @@ impl DbInner {
         tokio_handle: &Handle,
     ) -> Option<tokio::task::JoinHandle<()>> {
         let this = Arc::clone(self);
+        let mut is_shutdown = false;
         Some(tokio_handle.spawn(async move {
             let mut flusher = MemtableFlusher {
                 db_inner: this.clone(),
                 manifest,
             };
             let mut manifest_poll_interval =
                 tokio::time::interval(this.options.manifest_poll_interval);
-            loop {
+
+            // Stop the loop when the shut down has been received *and* all
+            // remaining `rx` flushes have been drained.
+            while !(is_shutdown && rx.is_empty()) {
                 tokio::select! {
                     _ = manifest_poll_interval.tick() => {
                         if let Err(err) = flusher.load_manifest().await {
```

Comment:
shouldn't `load_manifest` check for `is_shutdown`?

Reply:
Similar question to above: where do you think is the best place to put this? Unlike above, the load_manifest here is `async`. I'm not totally clear on where we can safely check.

Reply:
we can just not call `load_manifest` if `is_shutdown` is true right?

Reply:
Isn't there a race condition where `is_shutdown` is false, `load_manifest` is called, which calls:

```rust
        let current_manifest = self.manifest.refresh().await?;
```

At this await, the call stack is paused, then select is run again, which receives a shutdown message and sets `is_shutdown` to true. Then the `await` on the manifest refresh wakes up and continues.

I could be misunderstanding how the async/await runtime behaves. I'm new to async/await programming. 😬

Reply:
No, this can't happen. The call stack is paused at the call to `await`, so select can't run again until the stack resumes, exits `load_manifest`, moves to the next iteration of the loop and enters `select` again. Select won't spawn a new task for every branch that is ready, it just blocks the current task until one of the branches is ready, and then resumes the task.

Reply:
Gotcha! Well, that makes it easy. I'll do a PR to fix both.

Reply:
https://github.com/slatedb/slatedb/pull/138